src/compute/src/sink/continual_task.rs

```diff
@@ -0,0 +1,793 @@
+            // The known times at which we're going to write data to the output.
+            // This is guaranteed to include all times < append_times_progress,
+            // except that ones < output_progress may have been truncated.
+            append_times: BTreeSet::new(),
+            append_times_progress: Antichain::from_elem(Timestamp::minimum()),
+
+            // The data we've collected to append to the output. This is often
+            // compacted to advancing times and is expected to be ~empty in the
+            // steady state.
+            to_append: Vec::new(),
+            to_append_progress: Antichain::from_elem(Timestamp::minimum()),
+
+            // A lower bound on the upper of the output.
+            output_progress: Antichain::from_elem(Timestamp::minimum()),
+        }
+    }
+
+    /// Returns data to write to the output, if any, and the new upper to use.
+    ///
+    /// TODO(ct): Remove the Vec allocation here.
+    fn process(&mut self) -> Option<(Antichain<Timestamp>, Vec<((&K, &V), &Timestamp, &D)>)> {
+        // We can only append at times >= the output_progress, so pop off
+        // anything unnecessary.
+        while let Some(x) = self.append_times.first() {
+            if self.output_progress.less_equal(x) {
+                break;
+            }
+            self.append_times.pop_first();
+        }
+
+        // Find the smallest append_time before append_time_progress. This is
+        // the next time we might need to write data at. Note that we can only
+        // act on append_times once the progress has passed them, because they
+        // could come out of order.
+        let write_ts = match self.append_times.first() {
+            Some(x) if !self.append_times_progress.less_equal(x) => x,
+            Some(_) | None => {
+                // The CT sink's contract is that it only writes data at times
+                // we received an input diff. There are none in
+                // `[output_progress, append_times_progress)`, so we can go
+                // ahead and advance the upper of the output, if it's not
+                // already.
+                //
+                // We could instead ensure liveness by basing this off of
+                // to_append, but for any CTs reading the output (expected to be
+                // a common case) we'd end up looping each timestamp through
+                // persist one-by-one.
+                if PartialOrder::less_than(&self.output_progress, &self.append_times_progress) {
+                    return Some((self.append_times_progress.clone(), Vec::new()));
+                }
+                // Otherwise, nothing to do!
+                return None;
+            }
+        };
+
+        // We can also advance the output upper up to the write_ts if it's not
+        // there already.
+        if self.output_progress.less_than(write_ts) {
+            return Some((Antichain::from_elem(write_ts.clone()), Vec::new()));
+        }
```

Comment:
That reasoning makes sense, though I'm still wondering why this approach wouldn't work:

```
        if self.to_append_progress.less_equal(write_ts) {
            // Don't have all the necessary data yet.
            if self.output_progress.less_than(write_ts) {
                // We can advance the output upper up to the write_ts.
                // For self-referential CTs this might be necessary to ensure dataflow progress.
                return Some((Antichain::from_elem(write_ts.clone()), Vec::new()));
            }
            return None;
        }

        // Time to write some data!
        [...]
```

I.e., write the data immediately if we can't and only do the output frontier bumping if we can't yet. Probably doesn't matter too much for performance, but I also find this logic more intuitive.

Reply:
Oh, yeah, that's also fine. I double-checked with the unit tests and sqllogictests. I find both versions equally readable, so switched to yours.